src/GafferScene/MeshTangents.cpp

```diff
@@ -0,0 +1,155 @@
+	addChild( new StringPlug( "uvSet", Plug::In, "st" ) );
+	addChild( new StringPlug( "posPrimVarName", Plug::In, "P" ) );
+	addChild( new BoolPlug( "orthogonal", Plug::In, true ) );
+
+	// Fast pass-throughs for things we don't modify
+	outPlug()->attributesPlug()->setInput( inPlug()->attributesPlug() );
+	outPlug()->transformPlug()->setInput( inPlug()->transformPlug() );
+}
+
+MeshTangents::~MeshTangents()
+{
+}
+
+Gaffer::StringPlug *MeshTangents::uvSetPlug()
+{
+	return getChild<StringPlug>( g_firstPlugIndex );
+}
+
+const Gaffer::StringPlug *MeshTangents::uvSetPlug() const
+{
+	return getChild<StringPlug>( g_firstPlugIndex );
+}
+
+Gaffer::StringPlug *MeshTangents::posPrimVarNamePlug()
+{
+	return getChild<StringPlug>( g_firstPlugIndex + 1 );
+}
+
+const Gaffer::StringPlug *MeshTangents::posPrimVarNamePlug() const
+{
+	return getChild<StringPlug>( g_firstPlugIndex + 1 );
+}
+
+Gaffer::BoolPlug *MeshTangents::orthogonalPlug()
+{
+	return getChild<BoolPlug>( g_firstPlugIndex + 2 );
+}
+
+const Gaffer::BoolPlug *MeshTangents::orthogonalPlug() const
+{
+	return getChild<BoolPlug>( g_firstPlugIndex + 2 );
+}
+
+void MeshTangents::affects( const Gaffer::Plug *input, AffectedPlugsContainer &outputs ) const
+{
+	SceneElementProcessor::affects( input, outputs );
+
+	if( input == uvSetPlug() || input == posPrimVarNamePlug() || input == orthogonalPlug() )
+	{
+		outputs.push_back( outPlug()->objectPlug() );
+	}
+	else if( input == outPlug()->objectPlug() )
+	{
+		outputs.push_back( outPlug()->boundPlug() );
+	}
+}
+
+bool MeshTangents::processesBound() const
+{
+	return true;
+}
+
+void MeshTangents::hashProcessedBound( const ScenePath &path, const Gaffer::Context *context, IECore::MurmurHash &h ) const
+{
+	hashProcessedObject( path, context, h );
+}
+
+Imath::Box3f MeshTangents::computeProcessedBound( const ScenePath &path, const Gaffer::Context *context, const Imath::Box3f &inputBound ) const
+{
+	return inputBound;
+}
+
+bool MeshTangents::processesObject() const
+{
+	return true;
+}
+
+void MeshTangents::hashProcessedObject( const ScenePath &path, const Gaffer::Context *context, IECore::MurmurHash &h ) const
+{
+	uvSetPlug()->hash( h );
+	posPrimVarNamePlug()->hash( h );
+	orthogonalPlug()->hash( h );
+}
+
+IECore::ConstObjectPtr MeshTangents::computeProcessedObject( const ScenePath &path, const Gaffer::Context *context, IECore::ConstObjectPtr inputObject ) const
+{
+	const MeshPrimitive *mesh = runTimeCast<const MeshPrimitive>( inputObject.get() );
+	if( !mesh )
+	{
+		return inputObject;
+	}
+
+	std::string uvSet = uvSetPlug()->getValue();
+	std::string posPrimVarName = posPrimVarNamePlug()->getValue();
+	bool ortho = orthogonalPlug()->getValue();
+
+	std::pair<PrimitiveVariable, PrimitiveVariable> tangentPrimvars = MeshAlgo::calculateTangents( mesh, uvSet, ortho, posPrimVarName);
+	MeshPrimitivePtr meshWithTangents = runTimeCast<MeshPrimitive>( mesh->copy() );
+
+	meshWithTangents->variables["Tangent"] = tangentPrimvars.first;
```

Comment:
I wish you weren't so fast! I'm not convinced by this automatic naming approach, and I'd appreciate it if we discussed things before you changed Cortex APIs. I feel bad when I ask you to back up, so instead I guess I'm asking you not to charge ahead when there are still questions outstanding...

My reasoning is that having plugs for uTangent/vTangent names makes what the node does far more self-evident. The user can simply scan down the NodeEditor, and see that a specified uvSet will be used to generate a specified pair of primitive variables. This is much more obvious than trying to read the documentation. It's also way more obvious than having to do a ShufflePrimitiveVariables after the fact, after figuring out what it was those automatic names were.

I guess one argument for the automatic generation of names is that it gives us a standard that maybe allows downstream nodes to infer what they're getting, but this doesn't really fit with the general approach of primitive variables just being "along for the ride" (unless they're given specific meaning by the renderer backends). We don't know if the user will deform P after applying the primvars, and we don't even know what purpose they're being added for. So I think it's better to stick with the more lightweight API in Cortex, and give the user explicit control over the naming.

If you disagree, I'd love to hear your reasoning, but prior to code changes!